rfcs/0169-feature-parameter-names.md

```diff
@@ -0,0 +1,572 @@
+"withCaca"
+"withLua"
+"withQt"
+"withTeXLive"
+"withWxGTK"
+```
+
+I picked the `gnuplot` as example since it is the closest to be compliant with proposed rules.
+
+
+# Drawbacks
+
+1. The migration process involves renaming feature parameters, so these changes
+   will likely conflict with other outstanding changes to the derivation, and potentially
+   even with automatic version bumps.
+
+2. While migration is in the process (which is 1 year minimum due the grace
+   period), nixpkgs source will be even less consistent than it is today.
+
+3. Proposal migration plan slows down evaluation of all packages, including
+   ones without any feature parameters by approximately 1.3%
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Do nothing
+
+Avoids all the work and drawbacks, but there is no evidence that consistency
+problem will [solve itself evolutionary](https://github.com/NixOS/nixpkgs/pull/234463#issuecomment-1574892207).
+For some build dependencies, we have multiple equally popular feature parameter
+names, and people keep picking random one when adding new packages.
+
+## Do not address consistency issues across different features
+
+For most features, nixpkgs currently uses multiple names. Instead of process
+described in this RFC, we might rename parameters towards the most common one.
+That would mean
+
+```
+dbus => dbusSupport
+qt   => enableQt
+sdl  => withSDL  (winner by one usage)
+```
+
+## Other ways to do migration
+
+1. Just don't and call some release a flag day.
+
+2. Put deprecation logic into individual packages. That would avoid evaluation
+   penalty for already compliant packages and packages that has no feature
+   parameters, but would be quite verbose and require sophisticated automations
+   to introduce and drop.
+
+## Other ways to pass feature parameters
+
+### Group all feature parameters into separate attrset parameter
+
+Instead of writing
+
+```
+{ lib, stdenv, enable_foo ? true }:
+
+# uses enable_foo
+stdenv.mkDerivation { ... }
+```
+
+derivation can be written as
+```
+{ lib, stdenv, features ? { enable_foo = true; }}:
+
+# uses features.enable_foo
+stdenv.mkDerivation { ... }
+```
+
+It is definitely looks cleaner, but unfortunately hits several limitations of
+the Nix language.
+
+1. Nix language provides way to [introspect function argument names](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-functionArgs),
+   but no way to learn their default values. Learning default values is fundamentally impossible, since default values of one parameter might depend
+   on the value of another parameter, so default values can't be evaluated until function is called.
+
+2. Overrides become much more verbose. Simple and ubiquitous
+
+   ```
+   bar.override { with_foo = true; }
+   ```
+
+   becomes unwieldy
+
+   ```
+   bar.override (old: old // { features = old.features // { with_foo = true }; })
```

Comment:
you mean

```suggestion
   bar.overrideAttrs (old: old // { features = old.features // { with_foo = true }; })
```

?

Reply:
No, I mean `override`. If we do something like this:

```
{ lib, stdenv, features ? { with_foo = true; enable_bar = stdenv.isLinux; } }:

stdenv.mkDerivation { ... }
```

then we will have nice separation of feature flags and other stuff, but overrides become mess.